aws/resource_aws_sagemaker_device_fleet.go

```diff
@@ -0,0 +1,252 @@
+			},
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringLenBetween(1, 800),
+			},
+			"device_fleet_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.All(
+					validation.StringLenBetween(1, 63),
+					validation.StringMatch(regexp.MustCompile(`^[a-zA-Z0-9](-*[a-zA-Z0-9]){0,62}$`), "Valid characters are a-z, A-Z, 0-9, and - (hyphen)."),
+				),
+			},
+			"enable_iot_role_alias": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+			"iot_role_alias": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"output_config": {
+				Type:     schema.TypeList,
+				Required: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"kms_key_id": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validateArn,
+						},
+						"s3_output_location": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validation.StringLenBetween(1, 1024),
+						},
+					},
+				},
+			},
+			"role_arn": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateArn,
+			},
+			"tags":     tagsSchema(),
+			"tags_all": tagsSchemaComputed(),
+		},
+		CustomizeDiff: SetTagsDiff,
+	}
+}
+
+func resourceAwsSagemakerDeviceFleetCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
+	tags := defaultTagsConfig.MergeTags(keyvaluetags.New(d.Get("tags").(map[string]interface{})))
+
+	name := d.Get("device_fleet_name").(string)
+	input := &sagemaker.CreateDeviceFleetInput{
+		DeviceFleetName:    aws.String(name),
+		OutputConfig:       expandSagemakerFeatureDeviceFleetOutputConfig(d.Get("output_config").([]interface{})),
+		EnableIotRoleAlias: aws.Bool(d.Get("enable_iot_role_alias").(bool)),
+	}
+
+	if v, ok := d.GetOk("role_arn"); ok {
+		input.RoleArn = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	if len(tags) > 0 {
+		input.Tags = tags.IgnoreAws().SagemakerTags()
+	}
+
+	_, err := retryOnAwsCode("ValidationException", func() (interface{}, error) {
+		return conn.CreateDeviceFleet(input)
+	})
+	if err != nil {
+		return fmt.Errorf("error creating SageMaker Device Fleet %s: %w", name, err)
+	}
+
+	d.SetId(name)
+
+	return resourceAwsSagemakerDeviceFleetRead(d, meta)
+}
+
+func resourceAwsSagemakerDeviceFleetRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
+	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
+
+	deviceFleet, err := finder.DeviceFleetByName(conn, d.Id())
+	if err != nil {
+		if isAWSErr(err, "ValidationException", "No devicefleet with name") {
```

Comment:
Does the `ValidationException` have a nested error that has a specific code? We'd like to reduce the number of places we check the error message, since AWS recommends not relying on it.

Reply:
anyway i can verify this? (some common function the provider supports to check it)

Reply:
from a few tests, doesnt seem like it.